### PR TITLE
 Add dynamic form creation for participations 2.0

### LINKED ISSUE #16: Implement JSON schema form generation
Generate custom forms through a json form stored in an EventType

### PR DESCRIPTION
**Pull request by bbucsy**
Previous PR was good, but it can be better :point_right:
 - Added react_rails to render react components
 - With react components there is an intuitive GUI for building custom forms for admin users
 - There is a react component that renders the JSON schema to form
 - Reworked the workflow of event creation by adding a dropdown list of event_types, instead of the write id stuff
 
closes #16

### CODE REVIEW

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -14,10 +14,12 @@ def show
   # GET /events/new
   def new
     @event = Event.new
+    @event_types = EventType.all.pluck(:name, :id)
```
Comment (SepsiLaszlo) [inline app/controllers/events_controller.rb]: In dynamically typed languages we expect variables to contain what their names suggests.
In this instance we expect an array of `EventType` records, but we get an array of arrays which contains only two values of the model.
Theory aside, if I understand correctly, this is line is only needed to event form. So we could move this logic to the form partial.
In most cases we try to resist putting logic into views but partials can be exceptions. Partials meant to be reusable so the logic need for them to work should be provided.
I made the suggestion in the corresponding file.
```suggestion
```

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -14,10 +14,12 @@ def show
...
   end
 
   # GET /events/1/edit
   def edit
+    @event_types = EventType.all.pluck(:name, :id)
```
Comment (SepsiLaszlo) [inline app/controllers/events_controller.rb]: This also could be deleted.
```suggestion
```

File: app/views/events/_form.html.erb [hunk 3]
```diff
@@ -11,22 +11,36 @@
...
 
-  <div class="field">
-    <%= form.label :event_type_id %>
-    <%= form.number_field :event_type_id %>
+  <div class="input-group mb-12 mb-2">
+    <div class="input-group-prepend">
+      <%= form.label :event_type_id, "Jelentkező űrlap", class:"input-group-text" %>
+
+    </div>
+    <%= form.select :event_type_id, options_for_select(@event_types), include_blank: "Nincs jelentkezés" , class:"form-control" %>
```
Comment (SepsiLaszlo) [inline app/views/events/_form.html.erb]: This way the partial less context dependendent and more reusable.
```suggestion
    <%= form.select :event_type_id, options_for_select(EventType.all.pluck(:name, :id)), include_blank: "Nincs jelentkezés" , class:"form-control" %>
```
Maybe we could split this into multiple lines.

File: app/controllers/participations_controller.rb [hunk 4]
```diff
@@ -11,32 +11,42 @@ def index
...
   end
 
-  # GET /participations/new
+  # GET /events/1/register
   def new
-    @participation = Participation.new
+    event = Event.find(params[:event_id])
+    if event.event_type.nil?
+      redirect_to root_url, notice: "Ehhez az eseményhez nincs jelentkező űrlap"
+    elsif event.participations.exists?(:user => current_user)
```
Comment (SepsiLaszlo) [inline app/controllers/participations_controller.rb]: I think it is better to stick to newer hash synstax insted of the hash rocket, if possible
```suggestion
    elsif event.participations.exists?(user: current_user)
```

File: app/controllers/participations_controller.rb [hunk 5]
```diff
@@ -11,32 +11,42 @@ def index
...
+      redirect_to root_url, notice: "Erre az eseményre már regisztráltál"
+    else
+      @event_type    = event.event_type
+      @participation = Participation.new(:event => event)
```
Comment (SepsiLaszlo) [inline app/controllers/participations_controller.rb]: Here too:
```suggestion
      @participation = Participation.new(event: event)
```

File: db/seeds.rb [hunk 6]
```diff
@@ -5,3 +5,23 @@
...
+et = EventType.create(name:     "Demo form",
+                      schema:   '{"type":"object","title":"","properties":{"email":{"title":"E-mail","type":"string","format":"email"},"name":{"title":"Becenév","type":"string","default":"John Doe"},"id":{"title":"Személyi igazolvány szám","type":"string"},"glasses":{"enum":["igen","nem"],"title":"Szemüveges vagyok","type":"string"},"other":{"title":"Megjegyzés","type":"string"},"eula":{"title":"A szabályzatot elfogadom","type":"boolean"}},"dependencies":{},"required":["email","name","id","glasses","eula"]}',
+                      uischema: '{"glasses":{"ui:widget":"radio"},"other":{"ui:widget":"textarea"},"ui:order":["email","name","id","glasses","other","eula"]}')
+
+et_short = EventType.create(name:     "Short form",
+                            schema:   '{"type":"object","properties":{"yes":{"enum":["nem"],"title":"Igen?","type":"string"}},"dependencies":{},"required":["yes"],"title":"","description":""}',
+                            uischema: '{"yes":{"ui:widget":"radio"},"ui:order":["yes"]}')
+
+Event.create(post: p1, event_type: et)
+Event.create(post: p2, event_type: et_short)
```
Comment (SepsiLaszlo) [inline db/seeds.rb]: Please add the missing new line.

**Review by SepsiLaszlo — CHANGES_REQUESTED** [review]
Nice work!
The  custom form building page is really awesome!
I found some minor things to improve.

**Review by triszt4n — APPROVED** [review]

**Review by SepsiLaszlo — APPROVED** [review]

**Pull request merged**